pkg/events/redis/codec.go

```
@@ -0,0 +1,148 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"encoding/base64"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/events"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/unique"
+)
+
+func encodeEventData(evt events.Event) (string, error) {
+	pb, err := events.Proto(evt)
+	if err != nil {
+		return "", err
+	}
+	bpb, err := proto.Marshal(pb)
+	if err != nil {
+		return "", err
+	}
+	return protoEncodingPrefix + base64.StdEncoding.EncodeToString(bpb), nil
+}
+
+var errUnknownEncoding = errors.DefineInvalidArgument("unknown_encoding", "unknown encoding")
+
+func decodeEventData(enc string, evt *ttnpb.Event) error {
+	if !strings.HasPrefix(enc, protoEncodingPrefix) {
+		return errUnknownEncoding
+	}
+	bpb, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(enc, protoEncodingPrefix))
+	if err != nil {
+		return err
+	}
+	// NOTE: We're merging additional event data into an event that may already contain fields.
+	return proto.UnmarshalMerge(bpb, evt)
+}
+
+const (
+	eventUIDKey    = "event_uid"
+	eventNameKey   = "event_name"
+	entityIDKey    = "entity_id"
+	entityTypeKey  = "entity_type"
+	eventSparseKey = "event_sparse"
+)
+
+func encodeEventMeta(evt events.Event, id *ttnpb.EntityIdentifiers) []string {
```

I would prefer to propagate the errors here.